docs/api.md

```diff
@@ -969,6 +969,37 @@ I'm another middleware function
 Running myCommand!
 ```
 
+Middleware can be applied before validation by setting the second parameter to `true`.  This will execute the middleware prior to validation checks, but after parsing.
+
+Each callback is passed a reference to argv and context. The argv can be modified to affect the behavior of the validation and command execution.  The context contains the commands and availableOptions that matched the argv.  This feature is useful if you wish to populate a required option dynamically or based on the context prior to validation checks.
```

Comment:
I'm not opposed to adding `context` to middleware, but it feels like this feature should be a separate pull request? Sorry to be a pain, but let's keep this pull request focused on the ability to change the execution order of middleware.

Reply:
So, the need we had downstream was to take a required option that's on quite a few commands and populate it if its not provided and some other conditions are met.  Think of it like Heroku's CLI where they populate the value of the apps `-a` option on any command that requires it should the user be in the apps source folder.  

The only way for this to work is to be able to scan what command and options are possible, and compare it with what was passed in. The `innerArgv` alone doesn't give enough information to know if the option should be populated, or even if its available for the command.

Without the context pre-validation middleware doesn't really work for our circumstances.

Reply:
Don't you know this in the context of your middleware though?

If you know that you need a username/password, and your middleware receives an `argv` that does not have `argv.username` and `argv.password` you can take the step of populating the argument.

-----

The reason I'm pushing back is there are various places in the flow of yargs where it would be nice to have more context about the options that have been configured on the `yargs` object:

* what options are required?
* what top-level commands have been configured?
* etc.

I would like to put more thought into this feature rather than just serializing a couple parameters specific to your use-case.

I'd rather that the second argument to the middleware be the `yargs` object itself, and that we expand the API in yargs to better expose certain contextual information about the parsing. It's my opinion that this is a distinct feature from `pre-check-middleware`, however, and we should plan it out separately.

Reply:
part of the context you seem to need is what command the middleware is executing for?

`'commands': JSON.parse(JSON.stringify(context.commands))`

I think this would be better served by us figuring out an API surface that would let us provide pre-validation middleware to a specific command.

@evocateur @aorinevo any thoughts on this one; I was thinking maybe:

`[{handler: handler, applyBeforeValidation: true}]` <-- we could optionally make the handler passed for a command specific middleware be an object like this.